### ISSUE TITLE
Predicting with multi_output SoftmaxOutput

### ISSUE THREAD

**Issue by vchuravy** [dmlc/MXNet.jl#22]
I am working on image segmentation and I am training the following network to test out MXNet.

````julia
using MXNet

function create_conv(data, filters; kernel = (3, 3), pad=(1, 1), stride=(1,1))
  conv = @mx.chain mx.Convolution(data=data,
                                  kernel=kernel,
                                  num_filter=filters,
                                  pad=pad,
                                  stride=stride,
                                  workspace = 2300)  =>
                   mx.Activation(act_type=:relu)
end

function create_network()
  # Create network structure
  # Fully convolutional network
  data = mx.Variable(:data)
  conv1 = create_conv(data, 64)
  conv2 = create_conv(conv1, 128)
  conv3 = create_conv(conv2, 128)
  pool1 = mx.Pooling(data = conv3, kernel = (2,2), pool_type = (:max), stride = (2,2))
  conv4 = create_conv(pool1, 256)
  conv5 = create_conv(conv4, 512)
  conv6 = create_conv(conv5, 512)
  conv7 = create_conv(conv6, 2, pad = (0,0), kernel = (1, 1))
  deconv = mx.Deconvolution(data = conv7, num_filter = 2, kernel = (2,2), stride = (2,2))
  net = mx.SoftmaxOutput(data=deconv, name=:softmax, multi_output = true)
end
````

During training all seems fine:
````
INFO: Saved checkpoint to 'mx_2015-11-9_17-34/snapshots-0000.params'
INFO: == Epoch 001 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.4100
INFO:            time = 4059.5786 seconds
INFO: Saved checkpoint to 'mx_2015-11-9_17-34/snapshots-0001.params'
INFO: == Epoch 002 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.7100
INFO:            time = 4052.3061 seconds
INFO: Saved checkpoint to 'mx_2015-11-9_17-34/snapshots-0002.params'
INFO: == Epoch 003 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6600
INFO:            time = 4049.5518 seconds
INFO: Saved checkpoint to 'mx_2015-11-9_17-34/snapshots-0003.params'
````

When I then test the trained network on the training data to see how it worked out all the probabilities for the labels are equal (0.5 because right now I only have two classes).

My test script looks like this.
````
using HDF5
using MXNet

function load_test(filename)
  fid = h5open(filename, "r")

  data = readmmap(fid["data"])

  close(fid)
  mx.ArrayDataProvider(data, batch_size = 10)
end

model_dir = "./mx_2015-11-03_24-20"
prefix = "snapshots"
model_epoch = 4

model = mx.load_checkpoint(joinpath(model_dir, prefix), model_epoch, mx.FeedForward)

data_dir = "/data/pixel"
pred_dir = "/data/pixel/pred"
ids = ["100_train"]

run(`mkdir -p $pred_dir`)

for id in ids
  data = load_test(joinpath(data_dir, "$id.h5"))
  @time pred = mx.predict(model, data)
  h5open(joinpath(pred_dir, "$id.h5"), "w") do file
  write(file, "softmax", pred)
  end
end
````

Right now I am running prediction on a randomly initialized network to see if the problem prevails. 

Any ideas for different tests?

**Comment by pluskid** [comment]
While it is not hard to modify to support multiple output, the current built-in accuracy metric only works properly for single-output softmax predictions. Did you change that locally to make sure the monitoring values printed during training are actually meaningful?

**Comment by vchuravy** [comment]
No I did not.

But even running the initial network on the data only produces 0.5 as output.

**Comment by vchuravy** [comment]
See #23 for an attempt to calculate accuracy for a multi-output softmax

**Comment by pluskid** [comment]
Do you mean there might be a bug in prediction in the case of multi output softmax? You can use `model.arg_params` to get all the weights in the network. Probably you can assign some extreme weights to some layers (maybe the last layer, for example) and see if the output changes?

**Comment by vchuravy** [comment]
So in my network the output is defined by conv7 since it reduces the network to my 2 classes. Deconvolution afterwards is just for scaling.

Using the code below I set one of the kernel weights to 0 and the other one to 1. 

````
conv6_w = model.arg_params[symbol("mxnet.mx.convolution6_weight")]
mx.@nd_as_jl rw=conv6_w begin
         conv6_w[:, :, :, 1] = 0
         conv6_w[:, :,:, 2] = 1
end
````

The result from predict for every value is still 0.5

````
julia> minimum(pred2)
0.5f0

julia> maximum(pred2)
0.5f0
````

I am going to rewire the network and skip deconvolution and see what happens then.

**Comment by pluskid** [comment]
A side note: the weights in the predictor are **copied** from `arg_params` the first time `predict` is called. If you modified the weight, you need to pass `overwrite=true` to `predict` to tell it to re-copy the weights. I think it would be good to update the code so the weights in the predictor is sharing directly what is being trained, instead of copying.

**Comment by vchuravy** [comment]
OK then I will repeat my test again.

With the corrected accuracy measure my training looks a lot different. It seems to get stuck in a local optima.

````
INFO: Initializing parameters...
INFO: Creating KVStore...
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0000.params'
^[^[^[INFO: == Epoch 001 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6493
INFO:            time = 1105.7618 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0001.params'
INFO: == Epoch 002 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1095.0999 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0002.params'
INFO: == Epoch 003 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1096.7036 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0003.params'
INFO: == Epoch 004 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1101.8108 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0004.params'
INFO: == Epoch 005 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1097.7590 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0005.params'
INFO: == Epoch 006 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1100.4961 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0006.params'
INFO: == Epoch 007 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1095.9576 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0007.params'
INFO: == Epoch 008 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1096.0360 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0008.params'
INFO: == Epoch 009 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1096.2971 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0009.params'
INFO: == Epoch 010 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1096.1990 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0010.params'
INFO: == Epoch 011 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1097.0713 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0011.params'
INFO: == Epoch 012 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1096.9447 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0012.params'
INFO: == Epoch 013 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1100.5545 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0013.params'
INFO: == Epoch 014 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1092.0862 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0014.params'
INFO: == Epoch 015 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1091.4301 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0015.params'
INFO: == Epoch 016 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1091.9206 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0016.params'
INFO: == Epoch 017 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1092.3685 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0017.params'
INFO: == Epoch 018 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1091.8779 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0018.params'
INFO: == Epoch 019 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1091.8723 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0019.params'
INFO: == Epoch 020 ==========
INFO: ## Training summary
INFO:       :accuracy = 0.6690
INFO:            time = 1092.1349 seconds
INFO: Saved checkpoint to 'mx_2015-11-11_17-28/snapshots-0020.params'
````

**Comment by vchuravy** [comment]
So if I might venture a guess, multi_output softmax loses spatiality in the input dimensions.

So (64, 64, 2, 10) (w-h-kernels-samples) are reduced to (4096, 10). I deal with this in my code by reshaping the labels.

````
 labels = reshape(labels, *(size(labels, 1, 2, 3)...), size(labels, 4))
````

Now since C++ and Julia have different coloum-row orderings that loss of spatiality might mix things up.

**Comment by pluskid** [comment]
The Julia and C++ side use the same memory layout but different way of *interpreting* the shape. So reshaping will not mix anything. For example, reshaping an `NDArray` in Julia from (2,3,4) to (6, 4) is the same as reshaping this `NDArray` from (4,3,2) to (4, 6).
